### PR TITLE
os/drivers/lcd/: Add LCD logo support during LCD bootup

### PR DESCRIPTION
**Pull request by namanjain7**
LCD will be initialized with black screen

### CODE REVIEW

File: os/drivers/lcd/mipi_lcd.c [hunk 1]
```diff
@@ -236,6 +241,10 @@ static int lcd_putarea(FAR struct lcd_dev_s *dev, fb_coord_t row_start, fb_coord
 	priv->config->lcd_put_area((u8 *)lcd_buffer[lcd_buffer_index], row_start, col_start, row_end, col_end);
 	lcd_buffer_index = (1 - lcd_buffer_index);
 #else
+	if (lcd_data_logo != NULL) {
+		kmm_free(lcd_data_logo);
+		lcd_data_logo = NULL;
+	}
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:247]: Could you explain why this is added in this case only?

File: os/drivers/lcd/mipi_lcd.c [hunk 2]
```diff
@@ -399,6 +408,34 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
...
+		lcd_data_logo[lcd_data_index + 1] = lcd_logo_data[logo_arr_index++];
+		lcd_data_index += 2;
+		lcd_data_col_count += 1;
+		if (lcd_data_col_count == LCD_LOGO_XRES) {
+			lcd_data_index += ((CONFIG_LCD_XRES - LCD_LOGO_XRES) * 2);
+			lcd_data_col_count = 0;
+		}
+	}
+
+	priv->config->lcd_put_area((u8 *)lcd_data_logo, 1, 1, 480, 800);
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:436]: 480, 800 should be config_lcd_xres and yres.

File: os/drivers/lcd/lcd_logo_52_340.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+int LCD_LOGO_XRES = 52;
+int LCD_LOGO_YRES = 340;
+int LCD_LOGO_LOCATION = 221228;     //CONFIG_LCD_XRES * (CONFIG_LCD_YRES - LCD_LOGO_YRES) + (CONFIG_LCD_XRES - LCD_LOGO_XRES);
+
+int lcd_logo_data[52*340*2] = {0};   // Contains desired LOGO pixels
```
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/drivers/lcd/lcd_logo_52_340.c:28]: Is this printing which logo? just white? black? or something else?

[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/drivers/lcd/lcd_logo_52_340.c:28]: Could you use a definition for 0?
#define LCD_BLACK_VAL 0

File: os/drivers/lcd/mipi_lcd.c [hunk 4]
```diff
@@ -399,6 +408,34 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
...
 
+FAR void lcd_set_logo(FAR struct lcd_dev_s *dev)
+{
+	int logo_arr_index = 0;
+	int lcd_data_index = LCD_LOGO_LOCATION;
+	int lcd_data_col_count = 0;
+	FAR struct mipi_lcd_dev_s *priv = (FAR struct mipi_lcd_dev_s *)dev;
+
+#if !defined(CONFIG_LCD_SW_ROTATION)
+	lcd_data_logo = (uint8_t *)kmm_malloc(CONFIG_LCD_XRES * CONFIG_LCD_YRES * 2 + 1);
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:419]: Need to check null

File: os/drivers/lcd/mipi_lcd.c [hunk 5]
```diff
@@ -429,6 +467,7 @@ FAR struct lcd_dev_s *mipi_lcdinitialize(FAR struct mipi_dsi_device *dsi, struct
 		}
 		lcdvdbg("Memory allocated for SW screen rotation, Number of buffers created %d\n", NUM_OF_LCD_BUFFER);
 	}
+	lcd_data_logo = lcd_buffer[0];
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:470]: could you let me know what does this mean?

File: os/drivers/lcd/mipi_lcd.c (hunk 1 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c:247]: When ROTATION is enabled, then LCD driver already have buffer allocated for rotation purposes.
Since, in initial bootup (when displaying logo in screen), there is no put area call from application. So, we can safely use this rotation buffer to display our logo.
Now, when ROTATION is disabled, then we will need extra buffer which will be used for this specific logo display purpose. 
When application calls the putarea, there is no need of logo now, so this allocated buffer should be deleted.

File: os/drivers/lcd/mipi_lcd.c (hunk 5 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c:470]: There are two buffers, specifically used for rotation purposes.
In bootup time, there are no calls from application, so we can use the rotation buffer.
So, we are using one of the two buffer for logo display purposes.

File: os/drivers/lcd/Make.defs [hunk 6]
```diff
@@ -115,8 +115,16 @@ ifeq ($(CONFIG_LCD_ST7701SN),y)
   CSRCS += mipi_lcd.c
 endif
 
+ifeq ($(CONFIG_LCD_XRES),480)
+ifeq ($(CONFIG_LCD_YRES),800)
+CSRCS += lcd_logo_52_340.c
+endif
+endif
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/Make.defs:122]: Actually I name the file as lcd_logo_800_480 before, but Mr. Nam asked me to use 52_340 instead of lcd resolution.
https://github.ecodesamsung.com/TizenRT/Kernel/pull/666#discussion_r1315027

File: os/drivers/lcd/lcd_logo_52_340.c [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+int LCD_LOGO_XRES = 52;
+int LCD_LOGO_YRES = 340;
+int LCD_LOGO_LOCATION = 221228;     //CONFIG_LCD_XRES * (CONFIG_LCD_YRES - LCD_LOGO_YRES) + (CONFIG_LCD_XRES - LCD_LOGO_XRES);
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/lcd_logo_52_340.c:26]: To resolve this, I have added the logo resolution in kconfig.
And use it to calculate logo location in mipi_lcd.c file.

File: os/drivers/lcd/mipi_lcd.c (hunk 4 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c:416]: Added the null check

File: os/drivers/lcd/Make.defs (hunk 6 above)
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/drivers/lcd/Make.defs]: Because the file is a logo, I agree to get logo size instead of lcd size.
But, then, how will you add the same logo with different LCD size?

Is it possible to make proper logo buffer regarding to the size in runtime? or at build time?

File: os/drivers/lcd/mipi_lcd.c (hunk 1 above)
Comment (sunghan-chang) [reply os/drivers/lcd/mipi_lcd.c:244]: 1. Two buffers are necessary in ROTATION only? What about flushing thread? I mean when we need the flushing thread only without rotation, no need two buffers? Maybe not.
2. Could you leave some comment for that? It's difficult to know why it needs.

File: os/drivers/lcd/mipi_lcd.c [hunk 8]
```diff
@@ -356,7 +362,7 @@ static int lcd_init(FAR struct lcd_dev_s *dev)
 		lcddbg("ERROR: LCD Init sequence failed\n");
 	}
 	priv->config->init();
-	priv->config->lcd_enable();
+	priv->config->backlight(CONFIG_LCD_MAXPOWER);
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:365]: This looks not related with logo support. Could you let me know why you change lcd_enable to backlight control?

File: os/drivers/lcd/mipi_lcd.c [hunk 9]
```diff
@@ -399,6 +405,38 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
 	return OK;
 }
 
+FAR void lcd_set_logo(FAR struct lcd_dev_s *dev)
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:408]: How about lcd_put_logo or lcd_flush_logo or lcd_show_logo or something like that?
`set` looks just setting some value and looks flushing will be called later.

File: os/drivers/lcd/mipi_lcd.c [hunk 10]
```diff
@@ -399,6 +405,38 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
...
+#if !defined(CONFIG_LCD_SW_ROTATION)
+	lcd_data_logo = (uint8_t *)kmm_malloc(CONFIG_LCD_XRES * CONFIG_LCD_YRES * 2 + 1);
+	if (!lcd_data_logo) {
+		lcddbg("ERROR: LCD logo data memory allocation failed\n");
+		return;
+	}
+#endif
+	for (int i = 0; i < CONFIG_LCD_XRES * CONFIG_LCD_YRES * 2; i++) {
+		lcd_data_logo[i] = 0;
+	}
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:424]: Is this necessary?
The while loop (line 426) does not fill (set) data in every index?

File: os/drivers/lcd/mipi_lcd.c [hunk 11]
```diff
@@ -399,6 +405,38 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
...
+
+	while (logo_arr_index <= (CONFIG_LCD_LOGO_YRES * CONFIG_LCD_LOGO_XRES * 2)) {
+		lcd_data_logo[lcd_data_index] = lcd_logo_data[logo_arr_index++];
+		lcd_data_logo[lcd_data_index + 1] = lcd_logo_data[logo_arr_index++];
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:428]: lcd_data_logo is for sw rotation??

File: os/drivers/lcd/Make.defs (hunk 6 above)
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/Make.defs]: To resolve this issue, I have added a kconfig based on resolution of the logo. So, same logo can be used in different lcd as selected from menuconfig. https://github.com/Samsung/TizenRT/pull/6521/files#:~:text=CONFIG_LCD_LOGO_52_340)%2Cy)

And lcd buffer is created in runtime according to size of lcd. So, 480_800 lcd will create buffer accordingly. So, different lcd will be supported.
https://github.com/Samsung/TizenRT/pull/6521/files#diff-0dcbbbfee32bf5989cf8e2ad4117f850d4a44a336017e1f0e7f7e4ef484d5e2f:~:text=lcd_data_logo%20%3D%20(uint8_t%20*)kmm_malloc(CONFIG_LCD_XRES%20*%20CONFIG_LCD_YRES%20*%202%20%2B%201)%3B

File: os/drivers/lcd/mipi_lcd.c (hunk 2 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: done

File: os/drivers/lcd/lcd_logo_52_340.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/lcd_logo_52_340.c]: done

File: os/drivers/lcd/mipi_lcd.c (hunk 5 above)
Comment (sunghan-chang) [reply os/drivers/lcd/mipi_lcd.c:471]: Is there any reason why you set this here even you allocate new memory in the `lcd_set_logo` with `!defined(CONFIG_LCD_SW_ROTATION)`?
It's much difficult to analyse the code.. Let's keep the same location for the same purpose.

File: os/drivers/lcd/mipi_lcd.c [hunk 12]
```diff
@@ -38,11 +38,13 @@
 #define CONFIG_LCD_MAXPOWER 100
 #endif
 
+extern int lcd_logo_data[];
+static uint8_t *lcd_data_logo = NULL;
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c:42]: Is this below naming ok?
extern int lcd_logo_raw_data[];     // Buffer containing only logo
static uint8_t *lcd_logo_fullscreen_data = NULL;   // Buffer containing full screen data with logo on specific position

[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/drivers/lcd/mipi_lcd.c:42]: ok

**Comment by sunghan-chang** [comment]
@namanjain7 It takes too much time to get update. Could you do speed up?

File: os/drivers/lcd/mipi_lcd.c (hunk 11 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: I have formatted the code and added comment
https://github.com/Samsung/TizenRT/pull/6521/files#:~:text=/*%20Memory%20optimization%20applied,*/

Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: it's not used for sw rotation.
the rotation buffer is idle at the time of bootup. it's not being used. so, instead of creating and allocating new buffer of lcd size, we are just using the rotation buffer.

File: os/drivers/lcd/mipi_lcd.c (hunk 10 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: we are first filling the buffer with all 0's which is black color.
then add logo in specific index in buffer.

File: os/drivers/lcd/mipi_lcd.c (hunk 9 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: fixed.

File: os/drivers/lcd/mipi_lcd.c (hunk 5 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: I have added comment in code and allocated memory in lcd_put_logo function.

File: os/drivers/lcd/mipi_lcd.c (hunk 10 above)
Comment (sunghan-chang) [reply os/drivers/lcd/mipi_lcd.c]: Oh. ok. let's leave that description.

Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: I have added the description as comment in code.

File: os/drivers/lcd/Kconfig [hunk 13]
```diff
@@ -156,6 +156,30 @@ config LCD_YRES
...
+	int "LCD Logo X Resolution"
+	default 52 if LCD_ST7701SN
+	---help---
+		Specifies the X resolution of the LCD logo.
+
+config LCD_LOGO_YRES
+	int "LCD Logo Y Resolution"
+	default 340 if LCD_ST7701SN
+	---help---
+		Specifies the Y resolution of the LCD logo.
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/Kconfig]: I have removed the logo x and logo y from kconfig. instead user just selects the logo and then x and y will be taken from the file based on selection.

File: os/drivers/lcd/mipi_lcd.c [hunk 14]
```diff
@@ -399,6 +406,49 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
...
+	 * If rotation is enabled, then we have two buffers allocated for rotation.
+	 * During bootup, rotation buffer will not be used (No putarea call from application)
+	 * Therefore, the rotation buffer can be safely used for storing logo data.
+	 * 
+	 * If rotation is disabled, then we need to allocate memory for full screen data
+	 * and it will allocate memory to lcd_logo_fullscreen_data buffer.
+	 * 
+	 * If rotation is enabled, then lcd_logo_fullscreen_data contains pointer of rotation buffer */
+#if defined(CONFIG_LCD_SW_ROTATION)
+	lcd_logo_fullscreen_data = lcd_buffer[0];
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:426]: 1. maybe `lcd_buffer_index` instead of 0?
2. After flushing the log data with lcd_buffer[0], first ui data is using lcd_buffer[1] ?

File: os/drivers/lcd/mipi_lcd.c [hunk 15]
```diff
@@ -399,6 +406,49 @@ static int lcd_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
...
+	while (logo_arr_index < (LOGO_YRES * LOGO_XRES * 2)) {
+		lcd_logo_fullscreen_data[lcd_data_index] = lcd_logo_raw_data[logo_arr_index++];
+		lcd_logo_fullscreen_data[lcd_data_index + 1] = lcd_logo_raw_data[logo_arr_index++];
+		lcd_data_index += 2;
+		lcd_data_col_count += 1;
+		if (lcd_data_col_count == LOGO_XRES) {
+			lcd_data_index += ((CONFIG_LCD_XRES - LOGO_XRES) * 2);
+			lcd_data_col_count = 0;
+		}
+	}
```
Comment (sunghan-chang) [inline os/drivers/lcd/mipi_lcd.c:447]: Let's merge as it is. But logo pixels are overwritten. For performance, let's find a way to avoid overwritten.

Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c:447]: ok, I will find some algorithm to avoid overwritten.

File: os/drivers/lcd/mipi_lcd.c [hunk 16]
```diff
@@ -38,11 +39,13 @@
 #define CONFIG_LCD_MAXPOWER 100
 #endif
 
+extern uint8_t lcd_logo_raw_data[]; // Buffer containing only logo
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: fixed

File: os/drivers/lcd/lcd_logo_52_340.c [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
...,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/drivers/lcd/lcd_logo_52_340.c]: fixed

File: os/drivers/lcd/mipi_lcd.c (hunk 14 above)
Comment (namanjain7) [reply os/drivers/lcd/mipi_lcd.c]: updated to lcd_buffer[1]
now first ui data will use lcd_buffer[1]

File: os/drivers/lcd/lcd_logo_52_340.c [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <sys/types.h>
+#include "lcd_logo.h"
+
+#define LOGO_XRES 52
+#define LOGO_YRES 340
```
Comment (sunghan-chang) [inline os/drivers/lcd/lcd_logo_52_340.c:27]: Duplicated definitions (https://github.com/Samsung/TizenRT/pull/6521/files#diff-aecbe0525d3a8ed9962c2ebf8a044fa7c3d4072707efeee6495a1555796252e1R25-R26)

**Review by sunghan-chang — DISMISSED** [review]

File: os/drivers/lcd/lcd_logo.h [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __DRIVER_LCD_LOGO_H
+#define __DRIVER_LCD_LOGO_H
+
+#define LCD_BLACK_VAL 0
+
+#ifdef CONFIG_LCD_LOGO_52_340
```
Comment (sunghan-chang) [inline os/drivers/lcd/lcd_logo.h:24]: To use config, you should include <tinyara/config.h>.

**Review by sunghan-chang — APPROVED** [review]
